src/python/foglamp/core/scheduler.py

```diff
@@ -4,52 +4,419 @@
+                'scheduled_processes',
+                sa.MetaData(),
+                sa.Column('name', pg_types.VARCHAR(20)),
+                sa.Column('script', pg_types.JSONB))
+        # Class variables (end)
+
+        # Instance variables (begin)
+        self._start_time = None
+        """When the scheduler started"""
+        self.check_schedules_seconds = 5
+        """Frequency, in seconds, to check schedules to start tasks"""
+        self._paused = False
+        """When True, the scheduler will not start any new tasks"""
+        self._scheduled_processes = dict()
+        """Dictionary of scheduled_processes.id to script. Immutable."""
+        self._schedules = dict()
+        """Dictionary of schedules.id to _Schedule"""
+        self._schedule_executions = dict()
+        """Dictionary of schedules.id to _ScheduleExecution"""
+        self._processes = dict()
+        # pylint: disable=line-too-long
+        """Running processes
+
+        Dictionary of tasks.id to
+        `Process <https://docs.python.org/3/library/asyncio-subprocess.html#asyncio.asyncio.subprocess.Process>`_
+        """
+        # pylint: enable=line-too-long
+        # Instance variables (end)
+
+    async def stop(self):
+        """Attempts to stop the scheduler
+
+        Sends TERM signal to all running tasks. Does not wait for tasks to stop.
+
+        Prevents any new tasks from starting. This can be undone by setting the
+        _paused attribute to False.
+
+        Raises TimeoutError:
+            A task is still running. Wait and try again.
+        """
+        logging.getLogger(__name__).info("Stop requested")
+
+        self._paused = True
+
+        # Can not iterate over _processes - it can change mid-iteration
+        for key in list(self._processes.keys()):
+            try:
+                process = self._processes[key]
+            except KeyError:
+                continue
+
+            logging.getLogger(__name__).info(
+                    "Terminating pid %s for task %s",
+                    process.pid,
+                    key)
+
+            try:
+                process.send_signal(0)  # Check whether the process is running
+                process.terminate()
+            except (ProcessLookupError, OSError):
+                pass  # Process has terminated
+
+            await asyncio.sleep(.1)  # sleep 0 doesn't work
+
+        if self._processes:
+            raise TimeoutError()
+
+        # TODO: _schedule_executions should be empty
+        #if self._schedule_executions:
+        #    await asyncio.sleep(.1)  # sleep 0 doesn't work
+
+        self._start_time = None
+
+        return True
+
+    async def _start_startup_task(self, *args):
+        """Startup tasks are not tracked in the tasks table"""
+        task_id = str(uuid.uuid4())
+        logging.getLogger(__name__).info("Starting task %s %s", task_id, args)
+
+        # TODO: what if this fails?
+        process = await asyncio.create_subprocess_exec(*args)
```

Comment:
Log a fatal error in the short term. We should probably do better longer term. Looking at why it would not start, maybe retrying with a back-off etc. But to get the system running let's keep it simple.